### PR TITLE
4. API works with authInformation in properties

### LINKED ISSUE #1706: Composition service can create a workspace when no identity parameters are sent in the new workspace request.
Azure AD application registration created for a new workspace by the composition service when no identity parameters are sent in the new workspace request

- [x] Upgrade the workspace bundle AzureRM (#1714)
- [ ] Change the API to have a mandatory app_id with a value of auto if you want the bundle to create the aad part
- [x] Fix the API so it returns the correct error if app_id is missing #1705

### PR DESCRIPTION
**Pull request by ross-p-smith**
## What is being addressed

There was an authInformation block within the workspace defintion in cosmos. This is created by the API looking up these values from AAD. These need to move to workspace properties othewise when we automatically create the Workspace API App we will not be able to patch the workspace.

## How is this addressed

- Fix then dictionary check for app_id. If missing return AuthConfigValidationError
- Move the authInformation dictionary to the workspace properties
- if the app_id is `auto_create` set a property `register_aad_application` to be true. This will trigger the behaviour in the bundle.
- rename app_id to be client_id
- Pass the client_id, sp_id, app_role_id_* into Terraform so that Terraform can return them in outputs so that they dont get blanked out

Fixes #1706

### CODE REVIEW

File: e2e_tests/test_workspace_services.py [hunk 1]
```diff
@@ -8,6 +8,32 @@
...
+@pytest.mark.rossclean
+@pytest.mark.timeout(3000)
+async def test_ross_clean_workspace(admin_token, workspace_owner_token, verify) -> None:
+
+    workspace_paths = ["/workspaces/836dc1fd-635b-4bf0-82b0-4bd0eecc6408"]
+    for workspace_path in workspace_paths:
+        await disable_and_delete_resource(f'/api{workspace_path}', 'workspace', workspace_owner_token, admin_token, verify)
+
+
+@pytest.mark.ross
```
Comment (martinpeck) [inline e2e_tests/test_workspace_services.py:20]: ross?

File: e2e_tests/test_workspace_services.py [hunk 2]
```diff
@@ -8,6 +8,32 @@
 pytestmark = pytest.mark.asyncio
 
 
+@pytest.mark.rossclean
+@pytest.mark.timeout(3000)
+async def test_ross_clean_workspace(admin_token, workspace_owner_token, verify) -> None:
```
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply e2e_tests/test_workspace_services.py]: oh yes! thanks

File: api_app/models/schemas/workspace.py [hunk 3]
```diff
@@ -73,7 +72,7 @@ class Config:
                 "properties": {
                     "display_name": "the workspace display name",
                     "description": "workspace description",
-                    "app_id": "9d52b04f-89cf-47b4-868a-e12be7133b36",
+                    "client_id": "9d52b04f-89cf-47b4-868a-e12be7133b36",
```
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply api_app/models/schemas/workspace.py:75]: It represents a dummy value and is used to render the swagger page

File: e2e_tests/test_workspace_services.py [hunk 4]
```diff
@@ -8,6 +8,32 @@
 pytestmark = pytest.mark.asyncio
 
 
+@pytest.mark.rossclean
```
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply e2e_tests/test_workspace_services.py]: I've moved this out into a separate test file (it was how I was testing locally)

**Comment by damoodamoo** [comment]
Main question is what happens to existing resources - do we need a migration script to move the `authInformation` block into `properties` ? Aside from that I can't think of any other problems - and I like that we get the flexibility to modify via patches going forwards.

**Comment by ross-p-smith** [comment]
> Main question is what happens to existing resources - do we need a migration script to move the `authInformation` block into `properties` ? Aside from that I can't think of any other problems - and I like that we get the flexibility to modify via patches going forwards.

@damoodamoo  The migration script was written in a Previous PR and will be triggered based on a semantic version when I final settle on which version. Script is here: - /workspaces/AzureTRE/scripts/db_migrations.py. Method is called `moveAuthInformationToProperties`

File: api_app/db/repositories/workspaces.py [hunk 5]
```diff
@@ -63,22 +63,30 @@ def create_workspace_item(self, workspace_input: WorkspaceInCreate, auth_info: d
...
             properties=resource_spec_parameters,
-            authInformation=auth_info,
             resourcePath=f'/workspaces/{full_workspace_id}',
             etag=''  # need to validate the model
         )
 
         return workspace, template
 
+    def get_auto_app_registration_based_on_input(self, properties: dict) -> bool:
+        return "true" if properties["client_id"] == "auto_create" else "false"
```
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply api_app/db/repositories/workspaces.py:88]: Is it better to be explicit in the method name? (as above)?

[the remark this answers is not in the document]
Comment (martinpeck) [reply api_app/db/repositories/workspaces.py:88]: Looking at the name, it's not clear what gets returned.
If you make it explicit, you'd probably have a name that starts with  "is_" or "check_" at which point I'd expect it to return a bool, not a string.

Does the call site need a string? If it does, can it do the job to convert the bool?

[the remark this answers is not in the document]
Comment (martinpeck) [reply api_app/db/repositories/workspaces.py]: and that works?

**Comment by ross-p-smith** [comment]
/test-extended

(Just released my Resource Processor was on 0.1.12 instead of 0.1.18 - no wonder why it doesn't work!)

File: api_app/db/repositories/workspaces.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply api_app/db/repositories/workspaces.py]: yes

[the remark this answers is not in the document]
Comment (martinpeck) [reply api_app/db/repositories/workspaces.py]: Cool. Thanks for checking.

**Comment by ross-p-smith** [comment]
/test-extended

Tanya has now fixed the 3 failing E2E tests, so should be all green now

**Review by damoodamoo — APPROVED** [review]

**Pull request merged**